internal/services/storage/storage_account_resource.go

```diff
@@ -2462,29 +2462,60 @@ func expandArmStorageAccountAzureFilesAuthentication(input []interface{}) (*stor
 	v := input[0].(map[string]interface{})
 
 	directoryOption := storage.DirectoryServiceOptions(v["directory_type"].(string))
-	if _, ok := v["active_directory"]; directoryOption == storage.DirectoryServiceOptionsAD && !ok {
-		return nil, fmt.Errorf("`active_directory` is required when `directory_type` is `AD`")
-	}
 
+	var ad *storage.ActiveDirectoryProperties
+	switch string(directoryOption) {
+	case string(storage.DirectoryServiceOptionsAD):
+		if _, ok := v["active_directory"]; !ok {
+			return nil, fmt.Errorf("`active_directory` is required when `directory_type` is `AD`")
+		}
+		ad = expandArmStorageAccountActiveDirectoryProperties(v["active_directory"].([]interface{}))
+		if ad.AzureStorageSid == nil {
+			return nil, fmt.Errorf("`active_directory.0.storage_sid` is required when `directory_type` is `AD`")
+		}
+		if ad.DomainSid == nil {
+			return nil, fmt.Errorf("`active_directory.0.domain_sid` is required when `directory_type` is `AD`")
+		}
+		if ad.ForestName == nil {
+			return nil, fmt.Errorf("`active_directory.0.forest_name` is required when `directory_type` is `AD`")
+		}
+		if ad.NetBiosDomainName == nil {
+			return nil, fmt.Errorf("`active_directory.0.netbios_domain_name` is required when `directory_type` is `AD`")
+		}
+	case string(storageaccounts.DirectoryServiceOptionsAADKERB):
+		if _, ok := v["active_directory"]; ok {
+			ad = expandArmStorageAccountActiveDirectoryProperties(v["active_directory"].([]interface{}))
+		}
+	}
```

Comment:
Minor but we need to call the expand function for the AD properties in all cases so I think we can simplify this to
```suggestion
	ad := expandArmStorageAccountActiveDirectoryProperties(v["active_directory"].([]interface{}))
	if directoryOption == storage.DirectoryServiceOptionsAD {
		if ad == nil {
			return nil, fmt.Errorf("`active_directory` is required when `directory_type` is `AD`")
		}
		if ad.AzureStorageSid == nil {
			return nil, fmt.Errorf("`active_directory.0.storage_sid` is required when `directory_type` is `AD`")
		}
		if ad.DomainSid == nil {
			return nil, fmt.Errorf("`active_directory.0.domain_sid` is required when `directory_type` is `AD`")
		}
		if ad.ForestName == nil {
			return nil, fmt.Errorf("`active_directory.0.forest_name` is required when `directory_type` is `AD`")
		}
		if ad.NetBiosDomainName == nil {
			return nil, fmt.Errorf("`active_directory.0.netbios_domain_name` is required when `directory_type` is `AD`")
		}
	}
```

Reply:
I'm not sure I follow? I pulled your branch and ran the test with those changes and it appears to be fine?

Reply:
@magodo do you have an update?